src/main/java/reactor/core/publisher/MonoSingleEmitter.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2011-2016 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.core.publisher;
+
+import java.util.concurrent.atomic.*;
+import java.util.function.Consumer;
+
+import org.reactivestreams.*;
+
+import reactor.core.flow.Cancellation;
+import reactor.core.util.BackpressureUtils;
+import rx.exceptions.Exceptions;
+
+/**
+ * Wraps a the downstream Subscriber into a single emission object
+ * and calls the given callback to produce a signal (a)synchronously.
+ * @param <T> the value type
+ */
+public final class MonoSingleEmitter<T> extends Mono<T> {
+
+    final Consumer<SingleEmitter<T>> callback;
+
+    public MonoSingleEmitter(Consumer<SingleEmitter<T>> callback) {
+        this.callback = callback;
+    }
+
+    
+    @Override
+    public void subscribe(Subscriber<? super T> s) {
+        DefaultSingleEmitter<T> emitter = new DefaultSingleEmitter<>(s);
+        
+        s.onSubscribe(emitter);
+        
+        try {
+            callback.accept(emitter);
+        } catch (Throwable ex) {
+            Exceptions.throwIfFatal(ex);
+            emitter.error(ex);
+        }
+    }
+
+    static final class DefaultSingleEmitter<T> implements SingleEmitter<T>, Subscription {
```

Comment:
Because I need custom behavior in every method to clear the cancellation at the right time.